ui/versionmanager/version_downloader.go

```diff
@@ -0,0 +1,194 @@
+ */
+
+package versionmanager
+
+import (
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog/log"
+)
+
+type dlStatus string
+
+const (
+	inProgress dlStatus = "in_progress"
+	failed     dlStatus = "failed"
+	idle       dlStatus = "idle"
+	done       dlStatus = "done"
+)
+
+// Downloader node UI downloader
+type Downloader struct {
+	http *http.Client
+	lock sync.Mutex
+
+	progressUpdateEvery time.Duration
+
+	status     Status
+	statusLock sync.Mutex
+}
+
+// NewDownloader constructor for Downloader
+func NewDownloader() *Downloader {
+	return &Downloader{
+		http: &http.Client{
+			Timeout: time.Minute * 5,
+		},
+		status:              Status{Status: idle},
+		progressUpdateEvery: time.Second,
+	}
+}
+
+// DownloadOpts download options
+type DownloadOpts struct {
+	URL      *url.URL
+	DistFile string
+	Tag      string
+	Callback func(opts DownloadOpts) error
+}
+
+// DownloadNodeUI download node UI
+func (d *Downloader) DownloadNodeUI(opts DownloadOpts) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.status.Status == inProgress {
+		log.Warn().Msg("node UI download in progress - skipping")
+		return
+	}
+
+	go d.download(opts)
+}
+
+// Status download status
+// swagger:model DownloadStatus
+type Status struct {
+	Status      dlStatus `json:"status"`
+	ProgressPct int      `json:"progress_percent"`
+	Tag         string   `json:"tag,omitempty"`
+	Err         error    `json:"error,omitempty"`
+}
+
+func (s Status) withPct(p int) Status {
+	s.ProgressPct = p
+	return s
+}
+
+func (s Status) transition(status dlStatus) Status {
+	s.Status = status
+	return s
+}
+
+func (s Status) transitionWithErr(status dlStatus, err error) Status {
+	s.Status = status
+	s.Err = err
+	return s
+}
+
+// Status current download status
+func (d *Downloader) Status() Status {
+	return d.status
+}
+
+func (d *Downloader) download(opts DownloadOpts) {
+	d.update(Status{Status: inProgress, Tag: opts.Tag})
+
+	head, err := d.http.Head(opts.URL.String())
+	if err != nil {
+		d.update(d.status.transitionWithErr(failed, err))
+		return
+	}
+
+	totalSize, err := strconv.Atoi(head.Header.Get("Content-Length"))
+	if err != nil {
+		d.update(d.status.transitionWithErr(failed, err))
+		return
+	}
+
+	outFile, err := os.Create(opts.DistFile)
+	if err != nil {
+		d.update(d.status.transitionWithErr(failed, err))
+		return
+	}
+
+	doneChan := make(chan struct{})
+	go d.progressUpdater(doneChan, outFile, totalSize)
+
+	res, err := d.http.Get(opts.URL.String())
+	if err != nil {
+		d.update(d.status.transitionWithErr(failed, err))
+		return
+	}
+	defer res.Body.Close()
+
+	_, err = io.Copy(outFile, res.Body)
+	if err != nil {
+		d.update(d.status.transitionWithErr(failed, err))
+		return
+	}
+
+	close(doneChan)
+
+	if opts.Callback != nil {
+		log.Info().Msg("executing post download callback")
+		err := opts.Callback(opts)
+		if err != nil {
+			d.update(d.status.transitionWithErr(failed, err))
```

Comment:
Well the download didnt fail I think, just that callback, so is this really correct then?